### PR TITLE
Optimize Array.Clear using SpanHelpers

### PR DESCRIPTION
**Pull request by jkotas**
Reimplement most of Array.Clear in managed code using Span Clear helpers.

Fixes dotnet/corefx#29848

### CODE REVIEW

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline

**Comment by jkotas** [comment]
Array.Clear on empty array is 1.6x faster, 100 byte array is 3x faster, 10,000 byte array is 10x faster.

**Comment by stephentoub** [comment]
> Array.Clear on empty array is 1.6x faster, 100 byte array is 3x faster, 10,000 byte array is 10x faster.

That's great.  What about on an object array?

**Comment by jkotas** [comment]
Empty object array is 1.5x faster, 10 object element array is 3.2x faster, 1,000 object element array is 5.7x faster.

Small object arrays are about as fast as equivalent byte arrays. Large object arrays are slower than equivalent large byte arrays in the current implementation.

**Comment by stephentoub** [comment]
Cool, thanks.

**Comment by kasper3** [comment]
@jkotas, would it also have any impact on `.Fill()` method?

**Comment by jkotas** [comment]
> @jkotas, would it also have any impact on .Fill() method?

No.

File: src/System.Private.CoreLib/src/System/Array.cs [hunk 1]
```diff
@@ -265,8 +271,27 @@ public static void Copy(Array sourceArray, long sourceIndex, Array destinationAr
...
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        public static extern void Clear(Array array, int index, int length);
+        public static unsafe void Clear(Array array, int index, int length)
+        {
+            if (array == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+
+            ref byte p = ref GetRawArrayGeometry(array, out uint numComponents, out uint elementSize, out int lowerBound, out bool containsGCPointers);
+
+            if (index < lowerBound || index - lowerBound < 0 || length < 0 || (uint)(index - lowerBound + length) > numComponents)
```
Comment (ahsonkhan) [inline src/System.Private.CoreLib/src/System/Array.cs]: This is a relatively complex error check and I am having difficulty understanding it. Can you add a comment explaining it. Specifically, `(uint)(index - lowerBound + length) > numComponents`

File: src/System.Private.CoreLib/src/System/Array.cs [hunk 2]
```diff
@@ -265,8 +271,27 @@ public static void Copy(Array sourceArray, long sourceIndex, Array destinationAr
...
+            nuint byteLength = (uint)length * (nuint)elementSize;
+
+            if (containsGCPointers)
+                SpanHelpers.ClearWithReferences(ref Unsafe.As<byte, IntPtr>(ref start), byteLength / (uint)sizeof(IntPtr));
+            else
+                SpanHelpers.ClearWithoutReferences(ref start, byteLength);
+        }
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static extern ref byte GetRawArrayGeometry(Array array, out uint numComponents, out uint elementSize, out int lowerBound, out bool containsGCPointers);
```
[the remark this answers is not in the document]
Comment (kasper3) [reply src/System.Private.CoreLib/src/System/Array.cs]: Is there any issue tracking it?

File: src/System.Private.CoreLib/src/System/Array.cs (hunk 1 above)
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Array.cs]: I have added a helper local var to make it more self-explanatory.

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline

**Comment by Ilchert** [comment]
Hello, I found potential bug in this code.
In previous version method [`ZeroMemoryInGCHeap`](https://github.com/dotnet/coreclr/blob/074e9ca585d564ce48fefde744cc3c186aa8cbf1/src/vm/object.cpp#L1650) was used and it has special code to avoid `memset` compiler optimization (see bug [12207](https://github.com/dotnet/coreclr/issues/12207)). 
Now for big arrays method [`RuntimeImports.RhZeroMemory`](https://github.com/dotnet/coreclr/blob/f31097f14560b193e76a7b2e1e61af9870b5356b/src/System.Private.CoreLib/shared/System/SpanHelpers.cs#L337) is used which [uses](https://github.com/dotnet/coreclr/blob/2583ce936776a0eac31df904e41d5119840c203b/src/vm/comutilnative.cpp#L806) `memset`. 
I can not reprodure original bug, but this fix can have side effect.

**Comment by jkotas** [comment]
@Ilchert Thanks for pointing out the existing bug on this.

**Pull request merged**